mempool/src/tests/mocks.rs

```diff
@@ -22,36 +22,71 @@ use diem_types::{
 };
 use event_notifications::EventSubscriptionService;
 use futures::channel::{mpsc, oneshot};
-use mempool_notifications::{self, MempoolNotificationListener, MempoolNotifier};
+use mempool_notifications::{self, MempoolNotifier};
 use network::{
     peer_manager::{conn_notifs_channel, ConnectionRequestSender, PeerManagerRequestSender},
     protocols::network::{NewNetworkEvents, NewNetworkSender},
 };
 use std::sync::Arc;
 use storage_interface::{mock::MockDbReaderWriter, DbReaderWriter};
-use tokio::runtime::{Builder, Runtime};
+use tokio::runtime::{Builder, Handle, Runtime};
 use vm_validator::mocks::mock_vm_validator::MockVMValidator;
 
 /// Mock of a running instance of shared mempool.
 pub struct MockSharedMempool {
-    _runtime: Runtime,
+    _runtime: Option<Runtime>,
+    _handle: Option<Handle>,
     pub ac_client: mpsc::Sender<(SignedTransaction, oneshot::Sender<Result<SubmissionStatus>>)>,
     pub mempool: Arc<Mutex<CoreMempool>>,
     pub consensus_sender: mpsc::Sender<ConsensusRequest>,
-    pub mempool_notifier: Option<MempoolNotifier>,
+    pub mempool_notifier: MempoolNotifier,
 }
 
 impl MockSharedMempool {
     /// Creates a mock of a running instance of shared mempool.
     /// Returns the runtime on which the shared mempool is running
     /// and the channel through which shared mempool receives client events.
-    pub fn new(mempool_listener: Option<MempoolNotificationListener>) -> Self {
+    pub fn new() -> Self {
         let runtime = Builder::new_multi_thread()
             .thread_name("mock-shared-mem")
             .enable_all()
             .build()
             .expect("[mock shared mempool] failed to create runtime");
+        let (ac_client, mempool, consensus_sender, mempool_notifier) =
+            Self::start(runtime.handle());
+        Self {
+            _runtime: Some(runtime),
+            _handle: None,
+            ac_client,
+            mempool,
+            consensus_sender,
+            mempool_notifier,
+        }
+    }
+
+    /// Creates a mock of a running instance of shared mempool inside a tokio runtime;
+    /// Holds a runtime handle instead.
+    pub fn new_in_runtime() -> Self {
```

Comment:
It doesn't seem like this is currently used anywhere? I'm guessing this change is in preparation for a PR that will follow this one?

Reply:
I have a large PR for using this in new api test (when post /transactions to submit a new transaction), so split this out first as I also changed couple other places when doing the refactoring.

Reply:
Perfect, makes sense! Thanks @xli 😄